### PR TITLE
Check capture groups in queries

### PR DESCRIPTION
**Pull request by mattpolzin**
I borrowed and then heavily modified the exact script `nvim-treesitter` runs to validate the names of capture groups within the highlights file and applied that to CI for this repository so that we can make changes with confidence they will at least pass that portion of the `nvim-treesitter` project's CI.

It currently fails for a few values that I recall being problems the last time I submitted to the `nvim-treesitter` project (for example, `@punctuation` is not allowed without being qualified as `@punctuation.bracket`, `@punctuation.delimiter`, or `@punctuation.special` as indicated [here](https://github.com/nvim-treesitter/nvim-treesitter/blob/master/CONTRIBUTING.md#punctuation)).

These are easy enough failures to fix, but I wanted to get your thoughts on this addition to CI before fixing the capture names.

Checklist:

- [ ] All tests pass in CI.
- [x] There are sufficient tests for the new fix/feature.
- [x] Grammar rules have not been renamed unless absolutely necessary.
- [x] The conflicts section hasn't grown too much.
- [x] The parser size hasn't grown too much (check the value of STATE_COUNT in src/parser.c).

### CODE REVIEW

**Comment by lucario387** [comment]
I personally think the CI should not be run here, or if yes, needs to be limited to only query files that will be downstreamed to nvim-treesitter. This essentially locks out some probabilities of the parser to be integrated with other tree-sitter consumers, examples being Helix, Zed and Emacs.

**Comment by mattpolzin** [comment]
I modified the script I borrowed from `nvim-treesitter` to only check that the highlights.scm file avoids using `@` capture names that `nvim-treesitter` will reject during CI. My thinking was that these were capture names that I would need to change when bringing updates to `nvim-treesitter` and life would be easier if the highlights.scm file in this repository were more similar to the one in `nvim-treesitter`.

If the highlights.scm file benefits from being different in other downstream projects than it is in `nvim-treesitter`, then I agree with you that it is not appropriate to assert compatibility with `nvim-treesitter` in CI within this repo.

I think that one of three things could happen:
1. Merge this PR, keep capture names the same between this repository and `nvim-treesitter` until we know there's a reason not to. At that point, either remove this CI check or maybe have multiple versions of the highlights file in this repo if that is desirable (one of them could still be checked for compatibility with `nvim-treesitter`).
2. Make this check run but not actually fail -- it would provide information about incompatibilities, but fixing them would not be required to merge stuff into this repository.
3. Don't merge the PR. I just need to make more changes when porting things to `nvim-treesitter` and the highlights.scm file in this repo and the one in `nvim-treesitter` will diverge.

**Comment by clason** [comment]
> If the highlights.scm file benefits from being different in other downstream projects

This is indeed the case: every downstream project has its own set of captures (and directives/predicates!) We are actively working towards reducing the number of differences, but we will never be in complete alignment. So one size will never fit all.

Nevertheless, having _some_ queries in the repo is helpful to downstream, since it makes it much easier to look at _changes_ in the queries that were made in response to grammar changes (no matter what captures are used).

But even then, it only makes sense to test conformity with _upstream_ (tree-sitter); luckily, treesitter has recently made changes that will make this much easier in the upcoming release (`tree-sitter highlight --check`, for a minimal set of "canon captures", and an `--apply-all-captures` flag that reproduces Neovim's highlight strategy of "apply all matches, and let the last one win").

**TL;DR:** Do not use nvim-treesitter for parser testing (that is our job); stick to upstream tools (waiting for the next release if necessary).

**Comment by mattpolzin** [comment]
This tree-sitter business is a bigger ball of wax than I realized. 

I agree that this CI check is inappropriate given the above.

Thanks for the insights.

**Comment by clason** [comment]
> This tree-sitter business is a bigger ball of wax than I realized.

Oh, that's not even the half of it...
